contrib/mesos/pkg/proc/proc.go

```diff
@@ -0,0 +1,377 @@
+	defaultActionQueueDepth = 1024
+)
+
+type procImpl struct {
+	Config
+	backlog   chan Action    // action queue
+	terminate chan struct{}  // signaled via close()
+	wg        sync.WaitGroup // End() terminates when the wait is over
+	done      runtime.Signal
+	state     *stateType
+	pid       uint32
+	writeLock sync.Mutex    // avoid data race between write and close of backlog
+	changed   *sync.Cond    // wait/signal for backlog changes
+	engine    DoerFunc      // isolated this for easier unit testing later on
+	running   chan struct{} // closes once event loop processing starts
+	dead      chan struct{} // closes upon completion of process termination
+}
+
+type Config struct {
+	// cooldown period in between deferred action crashes
+	actionHandlerCrashDelay time.Duration
+
+	// determines the size of the deferred action backlog
+	actionQueueDepth uint32
+}
+
+var (
+	defaultConfig = Config{
+		actionHandlerCrashDelay: defaultActionHandlerCrashDelay,
+		actionQueueDepth:        defaultActionQueueDepth,
+	}
+	pid           uint32
+	closedErrChan <-chan error
+)
+
+func init() {
+	ch := make(chan error)
+	close(ch)
+	closedErrChan = ch
+}
+
+func New() Process {
+	return newConfigured(defaultConfig)
+}
+
+func newConfigured(config Config) Process {
+	state := stateNew
+	pi := &procImpl{
+		Config:    config,
+		backlog:   make(chan Action, config.actionQueueDepth),
+		terminate: make(chan struct{}),
+		state:     &state,
+		pid:       atomic.AddUint32(&pid, 1),
+		running:   make(chan struct{}),
+		dead:      make(chan struct{}),
+	}
+	pi.engine = DoerFunc(pi.doLater)
+	pi.changed = sync.NewCond(&pi.writeLock)
+	pi.wg.Add(1) // symmetrical to wg.Done() in End()
+	pi.done = pi.begin()
+	return pi
+}
+
+// returns a chan that closes upon termination of the action processing loop
+func (self *procImpl) Done() <-chan struct{} {
+	return self.done
+}
+
+func (self *procImpl) Running() <-chan struct{} {
+	return self.running
+}
+
+func (self *procImpl) begin() runtime.Signal {
+	if !self.state.transition(stateNew, stateRunning) {
+		panic(fmt.Errorf("failed to transition from New to Idle state"))
+	}
+	defer log.V(2).Infof("started process %d", self.pid)
+	var entered runtime.Latch
+	// execute actions on the backlog chan
+	return runtime.After(func() {
+		runtime.Until(func() {
+			if entered.Acquire() {
+				close(self.running)
+				self.wg.Add(1)
+			}
+			for action := range self.backlog {
+				select {
+				case <-self.terminate:
+					return
+				default:
+					// signal to indicate there's room in the backlog now
+					self.changed.Broadcast()
+					// rely on Until to handle action panics
+					action()
+				}
+			}
+		}, self.actionHandlerCrashDelay, self.terminate)
+	}).Then(func() {
+		log.V(2).Infof("finished processing action backlog for process %d", self.pid)
+		if !entered.Acquire() {
+			self.wg.Done()
+		}
+	})
+}
+
+// execute some action in the context of the current lifecycle. actions
```

Comment:
if "lifecycle" is not a widely-known term in this usage, please explain it. I'm not familiar with libprocess so "lifecycle" may have a well-known meaning in which case this is fine.